px4_ros2_cpp/include/px4_ros2/utils/frame_conversion.hpp

```diff
@@ -0,0 +1,180 @@
+/****************************************************************************
+ * Copyright (c) 2024 PX4 Development Team.
+ * SPDX-License-Identifier: BSD-3-Clause
+ ****************************************************************************/
+
+#pragma once
+
+#include <Eigen/Eigen>
+#include <px4_ros2/utils/geometry.hpp>
+
+namespace px4_ros2
+{
+
+/**
+ * @brief Converts attitude from NED to ENU frame.
+ * Performs reference frame change and quaternion rotation s.t. the identity quaternion points along the x-axis within its reference frame.
+ *
+ * @param q_ned Attitude quaternion in NED frame.
+ * @return Attitude quaternion in ENU frame.
+ */
+template<typename Type>
+Eigen::Quaternion<Type> attitudeNedToEnu(const Eigen::Quaternion<Type> & q_ned);
+
+/**
+ * @brief Converts attitude from ENU to NED frame.
+ * Performs reference frame change and quaternion rotation s.t. the identity quaternion points along the x-axis within its reference frame.
+ *
+ * @param q_enu Attitude quaternion in ENU frame.
+ * @return Attitude quaternion in NED frame.
+ */
+template<typename Type>
+Eigen::Quaternion<Type> attitudeEnuToNed(const Eigen::Quaternion<Type> & q_enu);
+
+/**
+ * @brief Transforms a point from body frame to world frame given the yaw of the body's attitude.
+ *
+ * @param yaw The yaw of the body's attitude [rad].
+ * @param point_body The point coordinates in the body frame.
+ * @return The yaw rotated point in the world frame.
+ */
+template<typename Type>
+Eigen::Matrix<Type, 3, 1> yawBodyToWorld(
+  Type yaw,
+  const Eigen::Matrix<Type, 3, 1> & point_body);
+
+/**
+ * @brief Converts yaw from NED to ENU frame.
+ *
+ * @param yaw_ned Yaw angle in NED frame [rad].
+ * @return Yaw angle in ENU frame [rad], wrapped to [-pi, pi].
+ */
+template<typename T>
+static inline T yawNedToEnu(const T yaw_ned_rad)
+{
+  return wrapPi(static_cast<T>(M_PI / 2.0) - yaw_ned_rad);
+}
+template float yawNedToEnu<float>(const float);
+template double yawNedToEnu<double>(const double);
```

Comment:
Is there a reason why you added these?

Reply:
No, I was confused thanks.
I was initially declaring the templates in the header and putting the template implementation and explicit types instantiation in the source file. Forgot to remove the type instantiating when moving inline implementations to the header.

But anyways as you mentioned I just learned that it's standard practice to have template methods be header only so I'll do that :+1: